File: Best Time to Buy and Sell Stock III/Best_Time_to_Buy_and_Sell_Stock_III.cpp
// 动态规划
// Runtime: 12 ms, faster than 15.16% of C++ online submissions for Best Time to Buy and Sell Stock III.
// Memory Usage: 5.7 MB, less than 2.29% of C++ online submissions for Best Time to Buy and Sell Stock III.

class Solution 
{
public:
    int maxProfit(vector<int>& prices) 
    {
        if (prices.size() < 2)
            return 0;

        int *buy = new int[3], *sell = new int[3]();
        buy[0] = 0;
        buy[1] = -prices[0];
        buy[2] = -prices[0];

        for (int i = 1; i < prices.size(); i++)
        {
            for (int j = 2; j >= 1; j--)
            {
                sell[j] = max(sell[j], buy[j] + prices[i]);
                buy[j] = max(buy[j], sell[j - 1] - prices[i]);
            }
        }
        int res = sell[2];

        delete[] sell;
        delete[] buy;

        return res;
    }
};

File: Best Time to Buy and Sell Stock III/Best_Time_to_Buy_and_Sell_Stock_III.py
# Runtime: 68 ms, faster than 59.03% of Python3 online submissions for Best Time to Buy and Sell Stock III.
# Memory Usage: 7.2 MB, less than 78.93% of Python3 online submissions for Best Time to Buy and Sell Stock III.
class Solution:
    def maxProfit(self, prices: 'List[int]') -> 'int':
        if len(prices) < 2:
            return 0
        buy = [-prices[0] for i in range(3)]
        buy[0] = 0
        sell = [0 for i in range(3)]

        for i in range(1, len(prices)):
            for j in range(2, 0, -1):
                sell[j] = max(sell[j], buy[j] + prices[i])
                buy[j] = max(buy[j], sell[j - 1] - prices[i])
        return sell[2]